internal/client.go

```diff
@@ -470,64 +484,97 @@ const (
 )
 
 // NewClient creates an instance of a workflow client
-func NewClient(service workflowservice.WorkflowServiceClient, domain string, options *ClientOptions) Client {
-	var identity string
-	if options == nil || options.Identity == "" {
-		identity = getWorkerIdentity("")
-	} else {
-		identity = options.Identity
+func NewClient(domain string, options ClientOptions) (Client, error) {
+	metricsScope := tagScope(options.MetricsScope, tagDomain, domain, clientImplHeaderName, clientImplHeaderValue)
+
+	if len(options.HostPort) == 0 {
+		options.HostPort = LocalHostPort
 	}
-	var metricScope tally.Scope
-	if options != nil {
-		metricScope = options.MetricsScope
+
+	if options.GRPCDialer == nil {
+		options.GRPCDialer = defaultGRPCDialer
 	}
-	metricScope = tagScope(metricScope, tagDomain, domain, clientImplHeaderName, clientImplHeaderValue)
-	var dataConverter DataConverter
-	if options != nil && options.DataConverter != nil {
-		dataConverter = options.DataConverter
-	} else {
-		dataConverter = getDefaultDataConverter()
+
+	connection, err := options.GRPCDialer(GRPCDialerParams{
+		HostPort:             options.HostPort,
+		RequiredInterceptors: requiredInterceptors(metricsScope),
+		DefaultServiceConfig: DefaultServiceConfig,
+	})
+
+	if err != nil {
+		return nil, err
 	}
-	var contextPropagators []ContextPropagator
-	if options != nil {
-		contextPropagators = options.ContextPropagators
+
+	return newServiceClient(workflowservice.NewWorkflowServiceClient(connection), connection, domain, options), nil
+}
+
+func newServiceClient(workflowServiceClient workflowservice.WorkflowServiceClient, connectionCloser io.Closer, domain string, options ClientOptions) Client {
```

Comment:
This signature is still required because it is used in all unit tests which pass service mock here.